### ISSUE TITLE
fatal runtime error from rustc if cranelift built proc-macro panics

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust#120075]
```console
> cargo new foo && cd foo
     Created binary (application) `foo` package
> echo '[lib]\nproc-macro=true' >> Cargo.toml
> echo '#[proc_macro] pub fn foo(_: proc_macro::TokenStream) -> proc_macro::TokenStream { panic!() }' > src/lib.rs
> echo 'fn main() { foo::foo!(); }' > src/main.rs
> CARGO_UNSTABLE_CODEGEN_BACKEND=true CARGO_PROFILE_DEV_CODEGEN_BACKEND=cranelift cargo build
   Compiling foo v0.1.0 (/tmp/scratch.rust-unwrapped.2024-01-17T23-01.KzoSlm/foo)
fatal runtime error: Rust cannot catch foreign exceptions
error: could not compile `foo` (bin "foo")

Caused by:
  process didn't exit successfully: `rustc --crate-name foo --edition=2021 src/main.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=272 --crate-type bin --emit=dep-info,link -C embed-bitcode=no -Z codegen-backend=cranelift -C debuginfo=2 -C metadata=6ab0c765ae937b1b -C extra-filename=-6ab0c765ae937b1b --out-dir /run/user/1000/cargo-home/target/shared/debug/deps -C incremental=/run/user/1000/cargo-home/target/shared/debug/incremental -L dependency=/run/user/1000/cargo-home/target/shared/debug/deps --extern foo=/run/user/1000/cargo-home/target/shared/debug/deps/libfoo-395c1bdc2cdecf24.so` (signal: 6, SIGABRT: process abort signal)
```

I expected to see this happen: A nice error message like an llvm-built proc-macro gives:

```console
> CARGO_UNSTABLE_CODEGEN_BACKEND=true CARGO_PROFILE_DEV_CODEGEN_BACKEND=llvm cargo build
   Compiling foo v0.1.0 (/tmp/scratch.rust-unwrapped.2024-01-17T23-01.KzoSlm/foo)
error: proc macro panicked
 --> src/main.rs:1:13
  |
1 | fn main() { foo::foo!(); }
  |             ^^^^^^^^^^^
  |
  = help: message: explicit panic

error: could not compile `foo` (bin "foo") due to 1 previous error
```

### Meta

`rustc --version --verbose`:
```
rustc 1.77.0-nightly (098d4fd74 2024-01-16)
binary: rustc
commit-hash: 098d4fd74c078b12bfc2e9438a2a04bc18b393bc
commit-date: 2024-01-16
host: x86_64-unknown-linux-gnu
release: 1.77.0-nightly
LLVM version: 17.0.6
```

**Comment by Nilstrieb** [comment]
panic=abort aborts the process, which is kind of expected. With the current architecture, I don't think there's much that can be fixed here. It would of course be nice if this worked.

**Comment by Nemo157** [comment]
`panic=abort` gives a different error message, the same with both llvm and cranelift, no mention of foreign exceptions or fatal runtime errors, just a clean `SIGABRT`:

```console
> CARGO_UNSTABLE_CODEGEN_BACKEND=true CARGO_PROFILE_DEV_CODEGEN_BACKEND=cranelift RUSTFLAGS=-Cpanic=abort cargo build
   Compiling foo v0.1.0 (/tmp/scratch.rust-unwrapped.2024-01-18T23-10.yyvQeV/foo)
warning: building proc macro crate with `panic=abort` may crash the compiler should the proc-macro panic

warning: `foo` (lib) generated 1 warning
error: could not compile `foo` (bin "foo")

Caused by:
  process didn't exit successfully: `rustc --crate-name foo --edition=2021 src/main.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=132 --crate-type bin --emit=dep-info,link -C embed-bitcode=no -Z codegen-backend=cranelift -C debuginfo=2 -C metadata=6ab0c765ae937b1b -C extra-filename=-6ab0c765ae937b1b --out-dir /run/user/1000/cargo-home/target/shared/debug/deps -C incremental=/run/user/1000/cargo-home/target/shared/debug/incremental -L dependency=/run/user/1000/cargo-home/target/shared/debug/deps --extern foo=/run/user/1000/cargo-home/target/shared/debug/deps/libfoo-395c1bdc2cdecf24.so -Cpanic=abort` (signal: 6, SIGABRT: process abort signal)
```

**Comment by bjorn3** [comment]
I am pretty sure this is because panic_unwind is linked rather than panic_abort and thus unwinding is attempted despite the necessary information for successful unwinding missing as cranelift doesn't support unwinding yet.
